src/boolean3.cpp

```diff
@@ -517,87 +421,36 @@ Boolean3::Boolean3(const Manifold::Impl &inP, const Manifold::Impl &inQ,
   p1q2_ = inQ_.EdgeCollisions(inP_);
   p2q1_ = inP_.EdgeCollisions(inQ_, true);  // inverted
 
-  p1q2_.Sort();
-  PRINT("p1q2 size = " << p1q2_.size());
-
-  p2q1_.Sort();
-  PRINT("p2q1 size = " << p2q1_.size());
-
   // Level 2
   // Find vertices that overlap faces in XY-projection
   SparseIndices p0q2 = inQ.VertexCollisionsZ(inP.vertPos_);
-  p0q2.Sort();
-  PRINT("p0q2 size = " << p0q2.size());
-
   SparseIndices p2q0 = inP.VertexCollisionsZ(inQ.vertPos_, true);  // inverted
-  p2q0.Sort();
-  PRINT("p2q0 size = " << p2q0.size());
-
-  if (3 * p1q2_.size() + 3 * p2q1_.size() > INT_MAX_SZ) {
-    valid = false;
-    return;
-  }
-
-  // Find involved edge pairs from Level 3
-  SparseIndices p1q1 = Filter11(inP_, inQ_, p1q2_, p2q1_);
-  PRINT("p1q1 size = " << p1q1.size());
 
 #ifdef MANIFOLD_DEBUG
   broad.Stop();
   Timer intersections;
   intersections.Start();
 #endif
 
-  // Level 2
-  // Build up XY-projection intersection of two edges, including the z-value for
-  // each edge, keeping only those whose intersection exists.
-  Vec<int> s11;
-  Vec<vec4> xyzz11;
-  std::tie(s11, xyzz11) = Shadow11(p1q1, inP, inQ, expandP_);
-  PRINT("s11 size = " << s11.size());
-
-  // Build up Z-projection of vertices onto triangles, keeping only those that
-  // fall inside the triangle.
-  Vec<int> s02;
-  Vec<double> z02;
-  std::tie(s02, z02) = Shadow02(inP, inQ, p0q2, true, expandP_);
-  PRINT("s02 size = " << s02.size());
-
-  Vec<int> s20;
-  Vec<double> z20;
-  std::tie(s20, z20) = Shadow02(inQ, inP, p2q0, false, expandP_);
-  PRINT("s20 size = " << s20.size());
-
   // Level 3
   // Build up the intersection of the edges and triangles, keeping only those
   // that intersect, and record the direction the edge is passing through the
   // triangle.
-  std::tie(x12_, v12_) =
-      Intersect12(inP, inQ, s02, p0q2, s11, p1q1, z02, xyzz11, p1q2_, true);
+  std::tie(x12_, v12_) = Intersect12(inP, inQ, p1q2_, expandP_, true);
   PRINT("x12 size = " << x12_.size());
 
-  std::tie(x21_, v21_) =
-      Intersect12(inQ, inP, s20, p2q0, s11, p1q1, z20, xyzz11, p2q1_, false);
+  std::tie(x21_, v21_) = Intersect12(inQ, inP, p2q1_, expandP_, false);
   PRINT("x21 size = " << x21_.size());
 
   if (x12_.size() > INT_MAX_SZ || x21_.size() > INT_MAX_SZ) {
     valid = false;
     return;
   }
 
-  s11.clear();
-  xyzz11.clear();
-  z02.clear();
-  z20.clear();
-
-  Vec<int> p0 = p0q2.Copy(false);
-  p0q2.Resize(0);
-  Vec<int> q0 = p2q0.Copy(true);
-  p2q0.Resize(0);
   // Sum up the winding numbers of all vertices.
-  w03_ = Winding03(inP, p0, s02, false);
+  w03_ = Winding03(inP, inQ, p0q2, expandP_, true);
```

Comment:
@elalish I'm thinking if it is possible to somehow integrate this winding number computation into `Intersect12`, but not sure if that is sound. Will it ever miss anything?

Reply:
Not sure if you already answered this for yourself, but w03 is calculated for every vertex. It doesn't need to be - I used to only calculate it for the verts of intersected edges and faces - you can flood fill the values to the rest because they're always consistent. I removed that because my flood-filling algorithm was slower than just calculating them all, but I may have just been doing it poorly.

Reply:
I think calculating them all is probably good for now, it is pretty fast.